### PR TITLE
bump MongoDB version to 4.2.1 (#253) and added support for updates with aggregation pipelines (#236)

### PR DESCRIPTION
**Pull request by bfreuden**
First proposal for issue/enhancement #236.
Second tentative since PR #251 has validation issues.
If successful, this PR replaces PR #251.

### CODE REVIEW

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 1]
```diff
@@ -157,6 +157,21 @@ MongoClient updateCollection(String collection, JsonObject query, JsonObject upd
 
   Future<@Nullable MongoClientUpdateResult> updateCollection(String collection, JsonObject query, JsonObject update);
 
+  /**
+   * Use an aggregation pipeline to update documents in the specified collection and return the handler with MongoClientUpdateResult result
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:161]: use `{@link MongoClientUpdateResult}`

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 2]
```diff
@@ -157,6 +157,21 @@ MongoClient updateCollection(String collection, JsonObject query, JsonObject upd
...
+   * @param collection  the collection
+   * @param query  query used to match the documents
+   * @param update used to describe how the documents will be updated
+   * @param resultHandler will be called when complete
+   */
+  @Fluent
+  MongoClient updateCollection(String collection, JsonObject query, JsonArray update,
+                               Handler<AsyncResult<@Nullable MongoClientUpdateResult>> resultHandler);
+
+  Future<@Nullable MongoClientUpdateResult> updateCollection(String collection, JsonObject query, JsonArray update);
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:172]: missing doc

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 3]
```diff
@@ -172,6 +187,21 @@ MongoClient updateCollectionWithOptions(String collection, JsonObject query, Jso
 
   Future<@Nullable MongoClientUpdateResult> updateCollectionWithOptions(String collection, JsonObject query, JsonObject update, UpdateOptions options);
 
+  /**
+   * Use an aggregation pipeline to update documents in the specified collection, specifying options and return the handler with MongoClientUpdateResult result
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:191]: use `{@link MongoClientUpdateResult}`

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 4]
```diff
@@ -172,6 +187,21 @@ MongoClient updateCollectionWithOptions(String collection, JsonObject query, Jso
...
+   * @param query  query used to match the documents
+   * @param update aggregation pipeline used to describe how documents will be updated
+   * @param options options to configure the update
+   * @param resultHandler will be called when complete
+   */
+  @Fluent
+  MongoClient updateCollectionWithOptions(String collection, JsonObject query, JsonArray update, UpdateOptions options,
+                                          Handler<AsyncResult<@Nullable MongoClientUpdateResult>> resultHandler);
+
+  Future<@Nullable MongoClientUpdateResult> updateCollectionWithOptions(String collection, JsonObject query, JsonArray update, UpdateOptions options);
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:203]: missing doc

**Comment by bfreuden** [comment]
@vietj thank you for the review! I have pushed a commit with javadoc changes you mentioned (and others as well).

**Comment by bfreuden** [comment]
@vietj If you had a little time to review my changes, that would be great :+1:. Otherwise no worries.

**Comment by bfreuden** [comment]
Hi all, 
I'm very sorry to write this message. I hope it won't bother you. 
Is there anything I have to do? Should I worry about something?

File: src/main/java/io/vertx/ext/mongo/MongoClient.java (hunk 3 above)
Comment (vietj) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java:212]: ping

File: src/main/java/io/vertx/ext/mongo/MongoClient.java (hunk 4 above)
Comment (vietj) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java:227]: ping

File: src/main/java/io/vertx/ext/mongo/MongoClient.java (hunk 3 above)
Comment (bfreuden) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java:212]: You'll have to be more specific.
I've already added the ``{@link MongoClientUpdateResult}``  (commit https://github.com/vert-x3/vertx-mongo-client/pull/254/commits/c7abe3ae5781f4107aa239c057c94ac1459293c2).
Is there something else?

File: src/main/java/io/vertx/ext/mongo/MongoClient.java (hunk 4 above)
Comment (bfreuden) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java:227]: You'll have to be more specific.
I've already added the following javadoc (commit https://github.com/vert-x3/vertx-mongo-client/pull/254/commits/c7abe3ae5781f4107aa239c057c94ac1459293c2):
```java
  /**
   * Like {@link #updateCollectionWithOptions(String, JsonObject, JsonArray, UpdateOptions, Handler)} but returns a {@code Future} of the asynchronous result
   */
```
Is there something else?

Comment (vietj) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java:227]: my bad, I did not spot it

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 5]
```diff
@@ -149,40 +161,85 @@ static MongoClient createWithMongoSettings(Vertx vertx, JsonObject config, Strin
...
   MongoClient updateCollectionWithOptions(String collection, JsonObject query, JsonObject update, UpdateOptions options,
                                           Handler<AsyncResult<@Nullable MongoClientUpdateResult>> resultHandler);
 
+  /**
+   * Like {@link #updateCollectionWithOptions(String, JsonObject, JsonObject, UpdateOptions, Handler)} but returns a {@code Future} of the asynchronous result
+   */
   Future<@Nullable MongoClientUpdateResult> updateCollectionWithOptions(String collection, JsonObject query, JsonObject update, UpdateOptions options);
 
+  /**
+   * Use an aggregation pipeline to update documents in the specified collection, specifying options and return the handler with MongoClientUpdateResult result
```
Comment (vietj) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:212]: `{@code MongoClientUpdateResult}`

**Comment by vietj** [comment]
sorry, the GitHub UI was still showing the previous result (it's an old tab), I just added comment for a last `@code` with MongoClientUpdateResult (or `@link` as you whish)

**Comment by bfreuden** [comment]
No problem!
It's done: ``{@code MongoClientUpdateResult}``
I also corrected a few missing ``{@code MongoClientDeleteResult}``

File: src/main/java/io/vertx/ext/mongo/MongoClient.java (hunk 5 above)
Comment (bfreuden) [reply src/main/java/io/vertx/ext/mongo/MongoClient.java]: done: {@code MongoClientUpdateResult}

**Review by vietj — APPROVED** [review]

**Comment by vietj** [comment]
thanks for the contribution

**Comment by bfreuden** [comment]
Thanks for your perseverance

**Pull request merged**